arrow-fx-coroutines/src/test/kotlin/arrow/fx/coroutines/stream/ops/CallbackTest.kt

```diff
@@ -0,0 +1,235 @@
+package arrow.fx.coroutines.stream.ops
+
+import arrow.fx.coroutines.Atomic
+import arrow.fx.coroutines.CancelToken
+import arrow.fx.coroutines.ExitCase
+import arrow.fx.coroutines.ForkAndForget
+import arrow.fx.coroutines.Promise
+import arrow.fx.coroutines.Schedule
+import arrow.fx.coroutines.StreamSpec
+import arrow.fx.coroutines.assertThrowable
+import arrow.fx.coroutines.guaranteeCase
+import arrow.fx.coroutines.milliseconds
+import arrow.fx.coroutines.never
+import arrow.fx.coroutines.parTupledN
+import arrow.fx.coroutines.seconds
+import arrow.fx.coroutines.sleep
+import arrow.fx.coroutines.stream.Chunk
+import arrow.fx.coroutines.stream.Stream
+import arrow.fx.coroutines.stream.cancellableCallback
+import arrow.fx.coroutines.stream.callback
+import arrow.fx.coroutines.stream.chunk
+import arrow.fx.coroutines.stream.compile
+import arrow.fx.coroutines.suspend
+import arrow.fx.coroutines.throwable
+import io.kotest.matchers.shouldBe
+import io.kotest.property.Arb
+import io.kotest.property.arbitrary.int
+import io.kotest.property.arbitrary.list
+import io.kotest.property.arbitrary.map
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.EmptyCoroutineContext
+import kotlin.coroutines.startCoroutine
+
+class CallbackTest : StreamSpec(iterations = 250, spec = {
```

Comment:
Should we add a test for when we emit elements after calling `end()`?

Reply:
Another potential test case could be if we emit from a separate coroutine. And also if we emit from a separate coroutine after the stream has ended.

Reply:
If you  emit after `end` it's offloaded in the `Queue` but never emitted downstream. Would you expect an exception here?

Reply:
Ideally, in the future we can have restrictions by the compiler. I started a chat about this on the KotlinLang Slack.

I guess throwing an exception at the calling site may make sense... As consumers have probably removed their subscription after the end signal. On other coroutine APIs they do emit and offer to differentiate if the consumer cares or not about the success of the push.

We could have a 3rd one that returns a result or either... Maybe we can have this as a separate ticket

Reply:
Moved here: https://github.com/arrow-kt/arrow-fx/issues/239